mmaction/core/evaluation/average_precision.py

```diff
@@ -0,0 +1,337 @@
+
+        gt = defaultdict(list)
+        for video_id, video in enumerate(videos):
+            # tube is a np.ndarray with (N, 5) shape,
+            # each row contains frame index and a bbounding box.
+            tubes = gt_tubes[video]
+
+            if label_index not in tubes:
+                continue
+
+            for tube in tubes[label_index]:
+                for t in tube:
+                    key = (video_id, int(t[0]))
+                    gt[key].append(t[1:5].tolist())
+
+        for key in gt:
+            gt[key] = np.array(gt[key].copy())
+
+        precision_recall = np.empty((det_result.shape[0] + 1, 2),
+                                    dtype=np.float32)
+        precision_recall[0, 0] = 1.0
+        precision_recall[0, 1] = 0.0
+        fn, fp, tp = sum([item.shape[0] for item in gt.values()]), 0, 0
+
+        for i, j in enumerate(np.argsort(-det_result[:, 3])):
+            key = (int(det_result[j, 0]), int(det_result[j, 1]))
+            box = det_result[j, 4:8]
+            is_positive = False
+
+            if key in gt:
+                ious = iou2d(gt[key], box)
+                max_idx = np.argmax(ious)
+
+                if ious[max_idx] >= threshold:
+                    is_positive = True
+                    gt[key] = np.delete(gt[key], max_idx, 0)
+
+                    if gt[key].size == 0:
+                        del gt[key]
+
+            if is_positive:
+                tp += 1
+                fn -= 1
+            else:
+                fp += 1
+
+            precision_recall[i + 1, 0] = tp / max(1, (tp + fp))
+            precision_recall[i + 1, 1] = tp / max(1, (tp + fn))
+
+        results.append(pr_to_ap(precision_recall))
+
+    frame_ap_result = np.mean(results * 100)
+    return frame_ap_result
+
+
+def frame_mean_ap_error(det_results, labels, videos, gt_tubes, threshold=0.5):
+    """Calculate error information for frame mAP in tubes.
+
+    The error information will contain ap_results, localization_error,
+    classification_error, time_error, other_error, missing_detections.
+
+    Args:
+        det_results (np.ndarray): Detection results for each frames.
+        labels (list): List of action labels.
+        videos (list): List of video names.
+        gt_tubes (dict): Ground truth tubes for each video. The format of
+            ``gt_tubes`` is {video_name: {label: list[tube]}}, where tube is a
+            np.ndarray with (N, 5) shape, each row contains frame index and a
+            bbounding box.
+        threshold (float): Threshold for IoU. Default: 0.5.
+
+    Returns:
+        dict: Result dict containing frame mAP, localization_error,
+            classification_error, time_error, other_error and
+            missing_detections
+    """
+    ap_results = []
+    other_ap_results = [[], [], [], []]
+    missing_detections = []
+    for label_index, label in enumerate(labels):
+        det_result = det_results[det_results[:, 2] == label_index, :]
```

Comment:
better to first divide det_results by label, or you need to enumerate det_results by n times instead of 1 time